src/auth/src/credentials/external_account.rs

```diff
@@ -467,4 +514,242 @@ mod test {
+                )))),
+                request::body(url_decoded(contains(("subject_token", "subject_token")))),
+                request::body(url_decoded(contains((
+                    "requested_token_type",
+                    ACCESS_TOKEN_TYPE
+                )))),
+                request::body(url_decoded(contains((
+                    "subject_token_type",
+                    JWT_TOKEN_TYPE
+                )))),
+                request::body(url_decoded(contains(("audience", "audience")))),
+                request::body(url_decoded(contains(("scope", DEFAULT_SCOPE)))),
+            ])
+            .respond_with(json_encoded(json!({
+                "access_token": "sts-only-token",
+                "issued_token_type": "urn:ietf:params:oauth:token-type:access_token",
+                "token_type": "Bearer",
+                "expires_in": 3600,
+            }))),
+        );
+
+        let creds = Builder::new(contents).build().unwrap();
+        let headers = creds.headers(Extensions::new()).await.unwrap();
+        match headers {
+            CacheableResource::New { data, .. } => {
+                let token = data.get("authorization").unwrap().to_str().unwrap();
+                assert_eq!(token, "Bearer sts-only-token");
+            }
+            CacheableResource::NotModified => panic!("Expected new headers"),
+        }
+    }
+
+    #[tokio::test]
+    async fn test_impersonation_flow_sts_call_fails() {
+        let subject_token_server = Server::run();
+        let sts_server = Server::run();
+        let impersonation_server = Server::run();
+
+        let impersonation_path = "/projects/-/serviceAccounts/sa@test.com:generateAccessToken";
+        let contents = json!({
+            "type": "external_account",
+            "audience": "audience",
+            "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+            "token_url": sts_server.url("/token").to_string(),
+            "service_account_impersonation_url": impersonation_server.url(impersonation_path).to_string(),
+            "credential_source": {
+                "url": subject_token_server.url("/subject_token").to_string(),
+                "format": {
+                  "type": "json",
+                  "subject_token_field_name": "access_token"
+                }
+            }
+        });
+
+        subject_token_server.expect(
+            Expectation::matching(request::method_path("GET", "/subject_token")).respond_with(
+                json_encoded(json!({
+                    "access_token": "subject_token",
+                })),
+            ),
+        );
+
+        sts_server.expect(
+            Expectation::matching(request::method_path("POST", "/token"))
+                .respond_with(status_code(500)),
+        );
+
+        let creds = Builder::new(contents).build().unwrap();
+        let err = creds.headers(Extensions::new()).await.unwrap_err();
+        assert!(err.is_transient());
```

Comment:
Q: does the error tell us which leg of the impersonated flow failed?

IMO it should, and we should be able to verify that programmatically.

Reply:
Ok, I see that one has a message: "failed to fetch token" and the other has a message "failed to exchange token".

And the underlying `reqwest::Error` might have a URL.

https://docs.rs/reqwest/latest/reqwest/struct.Error.html

Hm, this may be good enough.